modules/standard/CTypes.chpl

```diff
@@ -1012,15 +1035,15 @@ module CTypes {
     lifetime of the instance.  The returned pointer will be invalid if the
     instance is freed or even reallocated.
   */
-  inline proc c_ptrTo(ref c: class): c_ptr(void)
```

Comment:
This one was weird to track down and I'm not sure I understand it 100%, so here's what I got.

With `ref`, calls to `c_ptrTo` with an `unmanaged` actual were tripping [an assertion in code generation](https://github.com/chapel-lang/chapel/blob/1507af670379fdcc54af13d4a4b2f8ea2db4c8bc/compiler/codegen/cg-expr.cpp#L3589), _only_ in gasnet configuration. The assertion failure is during code generation of a `"cast"` primitive, and I assumed it was the one written in the body of `c_ptrTo`, but in the debugger I found it was actually a compiler-inserted one casting from (say) `Foo` to a `wide ref Foo`. (I figure this wouldn't be a `wide ref`, just a `ref` in single-locale configuration.)

With some more exploration I realized this happens when trying to use a `ref` to an `unmanaged` in gasnet config even without the cast to `c_ptr(void)` being in the equation, demonstrated by this example:
```
use CTypes;
class Foo { var x: int; }
var myunmanaged : unmanaged Foo = new unmanaged Foo();
﻿proc asdf(ref c: unmanaged) // same behavior with type `class` instead of `unmanaged`
{
  writeln(c); // ok
  writeln(c.borrow()); // assertion trips
  // writeln(__primitive("cast", c_ptr(void), c)); // assertion trips
  // writeln(__primitive("cast", c_ptr(void), c.borrow())); // assertion trips
}
asdf(myunmanaged);
```

Since the now-deprecated option of casting an `unmanaged` or anything else to `c_ptr(void)` worked, I was wondering what the difference was between that and the special `c_ptrTo` behavior. The cast was declared with two signatures, `inline operator :(x:borrowed, type t:c_ptr(void))` and `inline operator :(x:unmanaged, type t:c_ptr(void))` (which I think is why I previously thought we needed different logic for `unmanaged` classes). I think the relevant difference is that these use default argument intent, which I checked the spec to find is `const in` for `unmanaged`/`borrowed` and `const ref` for `owned`/`shared`. So it would have been taking `unmanaged`s as `const in`, and implicitly `borrow`ing `owned`/`shared` variables into a `const in` `borrowed`. I think using the default argument intent with a `class` type formal in `c_ptrTo` has the same effect as that did, except with the `borrow` explicit in the body instead.

I don't know exactly why the `unmanaged` `ref` failed in the way it did, but I'm thinking it's the wrong thing to be using (if it's valid at all) and in any case it's not what the old casts did.

Reply:
Gotcha, so by removing the `ref` it preserves the previous, correct behavior. And the `ref` being there sounds like it may have been a "bug waiting to happen", which we avoided by implementing `c_ptrTo` using the cast and not just the bare primitive.

Since this can be tripped independently of `c_ptr` casts (just calling `.borrow()` on a `ref unmanaged`), it may be worth making a separate issue for it as a bug. Either we should fix that or provide a better error than an assert.

Reply:
Yep that's my interpretation. And that sounds good to me, I created https://github.com/Cray/chapel-private/issues/5252 for this probably-bug.